cocotb/monitors/amba.py

```diff
@@ -0,0 +1,42 @@
+""" Copyright 2019 by Ben Coughlan <ben@liquidinstruments.com>
+
+Licensed under Revised BSD License
+All rights reserved. See LICENSE.
+"""
+
+import cocotb
+from cocotb.monitors import BusMonitor
+from cocotb.triggers import RisingEdge
+
+class AXI4(BusMonitor):
+    """AXI4 Bus
+
+    Just monitors for transaction on each channel, doesn't consider
+    transactions across multiple channels.
+    """
+    _signals = ["awvalid", "awready", "awaddr",
+                "wvalid", "wready", "wdata",
+                "bvalid", "bready", "bid",
+                "arvalid", "arready", "araddr",
+                "rvalid", "rready", "rdata"]
+    _optional_signals = []
+
+    @cocotb.coroutine
+    def _monitor_recv(self):
+        while True:
+            yield RisingEdge(self.clock)
+
+            if self.bus.awvalid.value and self.bus.awready.value:
+                self._recv(('awaddr', int(self.bus.awaddr)))
+
+            if self.bus.wvalid.value and self.bus.wready.value:
+                self._recv(('wdata', int(self.bus.wdata)))
+
+            if self.bus.bvalid.value and self.bus.bready.value:
+                self._recv(('bid', self.bus.bid.value))
+
+            if self.bus.arvalid.value and self.bus.arready.value:
+                self._recv(('araddr', int(self.bus.araddr)))
+
+            if self.bus.rvalid.value and self.bus.rready.value:
+                self._recv(('rdata', int(self.bus.rdata)))
```

Comment:
`rresp`?

Reply:
Possibly yes to both of these, not sure of the best format for them though
```python
self._recv(('rdata', int(self.bus.rdata))
self._recv(('rresp', int(self.bus.rresp))
```
or maybe change them all to 
```python
self._recv({
    'rdata': int(self.bus.rdata),
    'rresp': int(self.bus.rresp),
})
```
The dictionary option allows us to add all the qos, cache, prot, lock etc. as well.  It does make the test verification much more verbose though...

Reply:
> It does make the test verification much more verbose though

I'm not convinced.
```python
('rdata', 0xAA)
```
isn't really much better than
```python
{'rdata': 0xAA, 'rresp': 0}
```
or the alternate spelling
```python
dict(rdata=0xAA, rresp=0)
```